import dreem.util.util as util
from dreem.util.util import Argument
import click
import os
from click_option_group import optgroup


# Input files
out_dir = Argument('--out_dir', '-o', default=os.getcwd(), type=click.Path(exists=True), help='Where to output files')
sample = Argument('--sample', '-s', type=click.Path(exists=True), help='Name of the sequence alignment map file(s) folder')
fasta = Argument('--fasta', '-fa', type=click.Path(exists=True), help='Path to the fasta file', required=True)
fastq1 = Argument('--fastq1', '-fq1', help='Paths to the fastq1 file (forward primer). Enter multiple times for multiple files', type=click.Path(exists=True), required=True)
fastq2 = Argument('--fastq2', '-fq2', help='Paths to the fastq2 file (reverse primer). Enter multiple times for multiple files', type=click.Path(exists=True))
input_dir = Argument('--input_dir', '-id', type=click.Path(exists=True), help='Sequence alignment map files folder(s) generated by alignment', multiple=True)
samples = Argument('--samples', '-s', type=click.Path(exists=True), help='Path to the samples.csv file', default=None)
clustering_file = Argument('--clusters', '-cl', type=click.Path(exists=True), help='Path to the clustering.json file', default=None)
library = Argument('--library', '-l', type=click.Path(exists=True), help='Path to the library.csv file', default=None)

# Construct selection
coords = Argument('--coords', '-c', type=(str, int, int), multiple=True, help="coordinates for reference: '-c ref-name first last'")
primers = Argument('--primers', '-p', type=(str, str, str), multiple=True, help="primers for reference: '-p ref-name fwd rev'")
fill = Argument('--fill/--no-fill', None, type=bool, default=False, help="Fill in coordinates of reference sequences for which neither coordinates nor primers were given (default: no).")
parallel = Argument('--parallel', '-P', type=click.Choice(["profiles", "reads", "off", "auto"], case_sensitive=False), default="auto", help="Parallelize the processing of mutational PROFILES or READS within each profile, turn parallelization OFF, or AUTO matically choose the parallelization method (default: auto).")

# Demultiplexing
demultiplexing = Argument('--demultiplexing', '-dx', type=bool, help='Use demultiplexing', default=False)
barcode_start = Argument('--barcode_start', '-bs', type=int, help='Start position of the barcode in the read')
barcode_end = Argument('--barcode_end', '-be', type=int, help='End position of the barcode in the read')

# Bowtie2 TODO

# Cutadapt #TODO

# Clustering
clustering = Argument('--clustering', '-cl', type=bool, help='Use clustering', default=False)
n_clusters = Argument('--n_clusters', '-nc', type=int, help='Number of clusters', default=None)
max_clusters = Argument('--max_clusters', '-mc', type=int, help='Maximum number of clusters', default=None)
signal_thresh = Argument('--signal_thresh', '-st', type=float, help='Signal threshold', default=None)
info_thresh = Argument('--info_thresh', '-it', type=float, help='Information threshold', default=None)
include_g_u = Argument('--include_g_u', '-igu', type=bool, help='Include G and U', default=None)
include_del = Argument('--include_del', '-id', type=bool, help='Include deletions', default=None)
min_reads = Argument('--min_reads', '-mr', type=int, help='Minimum number of reads', default=None)
convergence_cutoff = Argument('--convergence_cutoff', '-cc', type=float, help='Convergence cutoff', default=None)
num_runs = Argument('--num_runs', '-nr', type=int, help='Number of runs', default=None)

# Aggregation
rnastructure_path = Argument('--rnastructure_path', '-rs', type=click.Path(exists=True), help='Path to RNAstructure, to predict structure and free energy', default=None)
rnastructure_temperature = Argument('--rnastructure_temperature', '-rst', type=bool, help='Use sample.csv temperature values for RNAstructure', default=False)
rnastructure_fold_args = Argument('--rnastructure_fold_args', '-rsa', type=str, help='Arguments to pass to RNAstructure fold', default=None)
rnastructure_dms = Argument('--rnastructure_dms', '-rsd', type=bool, help='Use the DMS signal to make predictions with RNAstructure', default=False)
rnastructure_dms_min_unpaired_value = Argument('--rnastructure_dms_min_unpaired_value', '-rsdmin', type=int, help='Minimum unpaired value for using the dms signal as an input for RNAstructure', default=0.01)
rnastructure_dms_max_paired_value = Argument('--rnastructure_dms_max_paired_value', '-rsdmax', type=int, help='Maximum paired value for using the dms signal as an input for RNAstructure', default=0.05)
rnastructure_partition = Argument('--rnastructure_partition', '-rspa', type=bool, help='Use RNAstructure partition function to predict free energy', default=False)
rnastructure_probability = Argument('--rnastructure_probability', '-rspr', type=bool, help='Use RNAstructure partition function to predict per-base mutation probability', default=False)
poisson = Argument('--poisson', '-po', type=bool, help='Predict Poisson confidence intervals', default=True)

# Misc
verbose = Argument('--verbose', '-v', type=bool, help='Verbose output', default=False)

@click.command()
@optgroup.group('Input files')
#@fastq1.cli()
#@fastq2.cli()
#@out_dir.cli()
#@library.cli()

@rnastructure_dms_max_paired_value.cli()

def cli(**args):
    print(args)

if __name__ == '__main__':
    cli()