# DREEM Vectoring Module
Contributor: Matty Allan, Yves Martin

## Purpose
- Filter out the reads with abnormaly high mutation rate (3 sigma interval)
- Convert mapped reads into mutation vectors


## Interface

### Input Files
- [=1] ```{reference}.fasta``` Sequence record file that contains the names and sequences of all references.
- [=1] ```library.csv``` CSV file containing the following columns:
```
- construct: name of the constructs (matching the ones of the fasta file).
- section_start: a 0-index of the beginning of the section.
- section_end: a 0-index of the end of the section (this index being non-included).
- section: optional name for the section. Default is {section_start}-{section_end}.

If you want to have several sections of the same construct, write the construct on two rows with different sections.
If you want the entire construct to be part of the output file, write the construct name without any section in addition of your sections rows.
```

- [≥1] ```/{input_dir}``` Sequence alignment map files folder(s) generated by ```alignment```. Its content is the following:
```bash
{input_dir}:= /path/to/{sample_1}/
  |- construct_1.bam
  |- construct_1.bam.bai
  |- construct_1_fastqc_report.txt
  |- construct_2.bam
  |- construct_2.bam.bai
  |- construct_2_fastqc_report.txt
  |- ...
{input_dir}:= /path/to/{sample_2}/
  |- ...
```

### Output Files
```bash
/{out_dir}/output/vectoring/{sample_1}/
  |- report.txt
  |- construct_1/
     |- section_1.orc
     |- section_2.orc
     |- ...
  |- construct_2.orc
    |- ...
  |- ...
/{out_dir}/output/vectoring/{sample_2}/
  |- ..
```

### Command-line usage
```dreem-vectoring -fa {reference}.fasta --out_dir [path] --input_dir [path/to/{sample_k}]```
- [=1] ```-fa / --fasta```: Input fasta file path.
- [≤1] ```-o / --out_dir```: output directory.
- [≥1] ```-id / --input_dir```: Sequence alignment map files folder(s) generated by ```alignment```. Last folder(s) of this (these) path(s) is `{sample_k}`, and `{sample_k}` will be used as a name for the output containing folder(s). `{input_dir}:= /path/to/{sample_k}`.
- [≥0] ```-c [ref | first | last]```: First and last coordinates for reference ref (creates one mutational profile); using 0 for last is shorthand for the last coordinate of the sequnce, -1 for the second-to-last, and so on.
- [≥0] ```-p [ref | fwd | rev]```: Forward and reverse primers for reference ref (creates one mutational profile)
- [≤1] ```--fill```: For every reference in ```reference.fasta``` that was not explicitly specified using a ```-c``` or ```-p``` option, create a mutational profile for the entire sequence. Note: if none of ```-c```, ```-p``` or ```--fill``` are given, then no mutational profiles will be generated.
- [≤1] ```-P```: Parallelize the processing of mutational ```profiles``` or process each profile in series and parallelize processing ```reads``` within each profile, turn all paralleization ```off```, or (default) ```auto```matically choose "reads" if processing 1 profile, otherwise "profiles".
