# DREEM Vectoring Module
Contributor: Matty Allan

## Functions
- Convert mapped reads into mutation vectors


## Interface

### Input Files
- {reference}.fasta [=1] Sequence record file that contains the names and sequences of all references.
- {sample}.bam      [≥1] Sequence alignment map file(s) generated by bowtie2 and "samtools view" during the alignment step. There must be one file for each sample, and its name must be "{sample_name}.bam". (DREEM will infer the sample name from the name of the BAM file and use that name for the mutation vector files.) The BAM file must have been generated using bowtie2 with the --xeq flag (to record matches and mismatches as = and X, respectively, instead of both as M). Ideally, the BAM file was generated by aligning reads to reference.fasta. If a different FASTA file is used, any additional or missing references will be ignored, and every reference that appears in both files must have exactly the same sequence in both files.
- {sample}.bam.bai  [≥1] An index for each BAM file(s) generated by "samtools index" during the alignment step. There must be one for each BAM file, located in the same directory as the corresponding BAM file and named "{sample_name}.bam.bai".

### Output Files
- mp/{reference}/{first}-{last}/{sample}_report.txt             [=1] Report containing the Sample Name, Reference Name, Reference Seq
- mp/{reference}/{first}-{last}/{sample}_vectors/batch_{≥0}.orc [≥0] Mutation vectors stored in Apache ORC format. If read-level parallelization is turned on, then one file is written by each batch process (the number is arbitrary); otherwise, a single file (batch_0.orc) is written.
