# DREEM Vectoring Module
Contributor: Matty Allan

## Purpose
- Convert mapped reads into mutation vectors


## Interface

### Input Files
- [=1] ```{reference}.fasta``` Sequence record file that contains the names and sequences of all references.
- [≥1] ```{sample}.bam``` Sequence alignment map file(s) generated by ```bowtie2``` and ```samtools view``` during the alignment step. There must be one file for each sample, and its name must be ```{sample_name}.bam```. (DREEM will infer the sample name from the name of the BAM file and use that name for the mutation vector files.) The BAM file must have been generated using ```bowtie2``` with the ```--xeq``` flag (to record matches and mismatches as ```=``` and ```X```, respectively, instead of both as ```M```). Ideally, the BAM file was generated by aligning reads to reference.fasta. If a different FASTA file is used, any additional or missing references will be ignored, and every reference that appears in both files must have exactly the same sequence in both files.
- [≥1] ```{sample}.bam.bai``` An index for each BAM file(s) generated by ```samtools index``` during the alignment step. There must be one for each BAM file, located in the same directory as the corresponding BAM file and named ```{sample_name}.bam.bai```.

### Output Files
- [=1] ```mp/{reference}/{first}-{last}/{sample}_report.txt``` Report containing the ```Sample Name```, ```Ref Name```, ```First``` and ```Last``` coordinates (inclusive) of the mutational profile, ```Ref Seq``` (the entire sequence, not just the region between the First and Last coordinates), ```Num Batches```, ```Num Vectors```, ```Speed (vec/s)```, ```Duration (s)```, time at which the processing ```Began``` and ```Ended```, and ```Checksums (md5)``` for all of the mutation vector files.
- [≥0] ```mp/{reference}/{first}-{last}/{sample}_vectors/batch_{≥0}.orc``` Mutation vectors stored in Apache ORC format. If read-level parallelization is turned on, then one file is written by each batch process (the number is arbitrary); otherwise, a single file (```batch_0.orc```) is written.

### Command-line usage
```mp-gen [-c ref first last]* [-p ref fwd rev]* [--fill] [-P profiles/reads/off/AUTO] reference.fasta sample.bam [sample2.bam ...]```
- ```mp-gen```: Wrapper for ```mp_gen``` function in ```mp.py```. Created by ```pip install --editable .``` using ```setup.py```.
- [≥0] ```-c ref first last```: First and last coordinates for reference ref (creates one mutational profile); using 0 for last is shorthand for the last coordinate of the sequnce, -1 for the second-to-last, and so on.
- [≥0] ```-p ref fwd rev```: Forward and reverse primers for reference ref (creates one mutational profile)
- [≤1] ```--fill```: For every reference in ```reference.fasta``` that was not explicitly specified using a ```-c``` or ```-p``` option, create a mutational profile for the entire sequence. Note: if none of ```-c```, ```-p``` or ```--fill``` are given, then no mutational profiles will be generated.
- [≤1] ```-P```: Parallelize the processing of mutational ```profiles``` or process each profile in series and parallelize processing ```reads``` within each profile, turn all paralleization ```off```, or (default) ```auto```matically choose "reads" if processing 1 profile, otherwise "profiles".
- [=1] ```reference.fasta```: The reference names and sequences.
- [≥1] ```sample.bam```: The aligned reads from the sample(s).
