# DREEM Vectoring Module
Contributor: Matty Allan, Yves Martin

## Purpose
- Filter out the reads with abnormaly high mutation rate (3 sigma interval)
- Convert mapped reads into mutation vectors


## Interface

### Input Files
- [=1] ```path/to/{reference}.fasta``` Sequence record file that contains the names and sequences of all references.
- [≥1] ```path/to/{sample}.bam``` Sequence alignment map file(s) generated by ```alignment```. Multiple files can be specified, and they do need to be in the same directory. Note that for every ```{sample}.bam``` file, a BAM index file with the same name (i.e. ```{sample}.bam.bai```) must be present in the same directory; this file is generated automatically near the end of ```alignment```. If the BAM index file is deleted or corrupted, it can be regenerated via the command ```samtools index {sample}.bam```.

### Output Files
- [≥1] ```{out_dir}/output/vectoring/{ref}/{first}-{last}/{sample}_report.txt``` Report in plain text format containing information about the vectoring of sample ```sample``` aligned to reference ```ref``` over the region from ```first``` to ```last```.
- [≥1] ```{out_dir}/output/vectoring/{ref}/{first}-{last}/{sample}/vectors_{num}.orc``` Mutation vectors in Apache ORC format. Each row is one mutation vector; each column is one position in the region from ```first``` to ```last```, inclusive (thus there are ```last - first + 1``` columns). Each element indicates the possible mutation states at that position in the region. Each element is one byte that is the logical union of all possible states for that position in the mutation vector:
  - ```00000001```: match
  - ```00000010```: deletion
  - ```00000100```: ≥1 base is inserted immediately 3' of this position
  - ```00001000```: ≥1 base is inserted immediately 5' of this position
  - ```00010000```: substitution to A
  - ```00100000```: substitution to C
  - ```01000000```: substitution to G
  - ```10000000```: substitution to T

Additionally, ```00000000``` indicates that the mutation vector does not cover the position. For example, ```00000011``` means that the element of the mutation vector could be a match or deletion, and ```11110001``` means a match or any substitution. Note several important differences between the encoding of insertions and deletions:
- Every base deleted from the read results in one byte with a deletion bit (unless the deletion is ambiguous). Every insertion results in two bytes with an insertion bit: a 3' insertion bit set immediately upstream of the insertion (unless it occurs at the first position in the region) and a 5' insertion bit immediately downstream (unless the insertion occurs the last position in the region).
- A deletion bit indicates that the position is deleted from the read. However, insertion bits indicate that one or more bases are inserted 5' or 3' of the position but do not indicate anything about the position itself. Thus, every position with an insertion bit also has one or more bits indicating the state of the position itself. For example, if the reference is ```GACTA``` and the read is ```GACATA```, then the ```C``` will be ```00000101``` (a match with an insertion on its 3' side) and the ```A``` will be ```00001001``` (a match with an insertion on its 5' side).
- For deletions, every base deleted from the read corresponds to one base in the reference, so the deleted bases (and read sequence) can be determined exactly given only the reference sequence and the mutation vector (unless the deletion is ambiguous). However, mutation vector files store only the location of an insertion, but not its length or sequence (as this information is not needed for DREEM and doing so would require inserting at least one extra byte to the mutation vector), so the inserted bases (and read sequence) cannot be determined given only the reference sequence and the mutation vector.

### Command-line usage
```dreem-vectoring -fa {reference}.fasta --out_dir [path] --input_dir [path/to/{sample_k}]```
- [=1] ```-fa / --fasta```: Input fasta file path.
- [≤1] ```-o / --out_dir```: output directory.
- [≥1] ```-id / --input_dir```: Sequence alignment map files folder(s) generated by ```alignment```. Last folder(s) of this (these) path(s) is `{sample_k}`, and `{sample_k}` will be used as a name for the output containing folder(s). `{input_dir}:= /path/to/{sample_k}`.
- [≥0] ```-c [ref | first | last]```: First and last coordinates for reference ref (creates one mutational profile); using 0 for last is shorthand for the last coordinate of the sequnce, -1 for the second-to-last, and so on.
- [≥0] ```-p [ref | fwd | rev]```: Forward and reverse primers for reference ref (creates one mutational profile)
- [≤1] ```--fill```: For every reference in ```reference.fasta``` that was not explicitly specified using a ```-c``` or ```-p``` option, create a mutational profile for the entire sequence. Note: if none of ```-c```, ```-p``` or ```--fill``` are given, then no mutational profiles will be generated.
- [≤1] ```-P```: Parallelize the processing of mutational ```profiles``` or process each profile in series and parallelize processing ```reads``` within each profile, turn all paralleization ```off```, or (default) ```auto```matically choose "reads" if processing 1 profile, otherwise "profiles".
