# DREEM Alignment Module
Contributor: Matty Allan

## Purpose
- Convert sequencing files into BAM files

## Interface

### Input Files
- [=1] ```{reference}.fasta```. Sequence record file that contains the names and sequences of all references.
- [≥1] ```{sample_x}_1.fastq```. Sequence alignment file(s) generated by the sequencer. There must be one file for each sample, and its name must be ```{sample_x}_1.fastq```.  
- [≥1] ```{sample_x}_2.fastq```. Sequence alignment file(s) generated by the sequencer. There must be one file for each sample, and its name must be ```{sample_x}_2.fastq```.  

### Output Files
- [≥1] ```{sample_x}_{construct_xy}.bam```. One BAM file per construct (sequence name) in each sample.  

### Command-line usage

```dreem-alignment --fastq1 [path to file] --fastq2 [path to file] --fasta [path to file]```

- ```dreem-alignment```: Wrapper for ```run``` function in ```dreem/alignment/run.py```. 
- [=1] `--fasta` : ```{reference}.fasta```
- [≥1] `--fastq1`: ```{sample_x}_1.fastq```
- [≥1] `--fastq2`: ```{sample_x}_2.fastq```
- [≥1] `output`: output repository.
- `[bowtie2 args]`
- `[cutadapt args]`
